### PR TITLE
#3 Nephew grand child

### PR DESCRIPTION
**Pull request by lamsecrettune**
Lam completed project

### CODE REVIEW

**Comment by vtd12** [comment]
Looks almost like mine @@

**Comment by vtd12** [comment]
I can say you follow the instruction in the document step by step strictly. The test cases are also quite comprehensive to me. I don't know if it's good or not but I think you commented too much lol.

**Comment by 0xCheeezzyyyy** [comment]
Since _balances and _allowances are declared public, no need for getter functions since they can be queried directly. Following the ERC20 standard getter functions, perhaps you could declare _balances and _allowances as private instead.

Burn function could include another variant to burn on behalf of someone else as long as the amount <= allowance given, ie. for other smart contract interaction haha just a suggestion.

Test cases seems really comprehensive, maybe you could classify the tests into different catergories by having more describe("Type of Test Category", () => <All tests belonging to this category>} to make it print more neater haha

File: contracts/ERC20.sol [hunk 1]
```diff
@@ -1,34 +1,112 @@
...
+  uint256 public constant MAXINT = 2**256 - 1;
+  uint256 private _totalSupply;
+  mapping(address => uint256) private _balances;
+  string private _coinName;
+  string private _coinSymbol;
+  uint8 private _coinDecimals;
+  mapping(address => mapping(address => uint256)) private _allowances;
+
+  constructor(uint256 _initialSupply, string memory _name, string memory _symbol, uint8 _decimals) public {
+    _coinName = _name;
```
Comment (UncleGrandpa925) [inline contracts/ERC20.sol:16]: you can just name it `_name`. `_coinName` sounds a bit odd haha

File: contracts/ERC20.sol [hunk 2]
```diff
@@ -1,34 +1,112 @@
...
+    _coinSymbol = _symbol;
+    _coinDecimals = _decimals;
+    mint(msg.sender, _initialSupply);
```
Comment (UncleGrandpa925) [inline contracts/ERC20.sol:19]: this is good

File: contracts/ERC20.sol [hunk 3]
```diff
@@ -1,34 +1,112 @@
...
 
-  function transfer(address to, uint256 amount) external returns (bool) {
+  function transfer(address to, uint256 amount) external override returns (bool) {
+    // Check if the amount is greater than the balance of the sender
+    require(_balances[msg.sender] >= amount, "Not enough balance");
+    // Check if the recipient is not the zero address
+    require(to != address(0), "Invalid recipient");
+    // Decrease the sender's balance
+    _balances[msg.sender] -= amount;
+    // Increase the recipient's balance
```
Comment (UncleGrandpa925) [inline contracts/ERC20.sol:47]: You have a bit too many comments 🤣 Good code is when there is no need to comment, or only comments for high-level ideas

**Pull request closed without merging**